js/plugins/vertexai/src/reranker.ts

```diff
@@ -0,0 +1,166 @@
+  RerankerAction,
+  rerankerRef,
+} from '@genkit-ai/ai/reranker';
+import { GoogleAuth } from 'google-auth-library';
+import z from 'zod';
+import { PluginOptions } from '.';
+
+const DEFAULT_MODEL = 'semantic-ranker-512@latest';
+
+const getRerankEndpoint = (projectId: string, location: string) => {
+  return `https://discoveryengine.googleapis.com/v1/projects/${projectId}/locations/${location}/rankingConfigs/default_ranking_config:rank`;
+};
+
+// Define the schema for the options used in the Vertex AI reranker
+export const VertexAIRerankerOptionsSchema = z.object({
+  k: z.number().optional().describe('Number of top documents to rerank'), // Optional: Number of documents to rerank
+  model: z.string().optional().describe('Model name for reranking'), // Optional: Model name, defaults to a pre-defined model
+  location: z
+    .string()
+    .optional()
+    .describe('Google Cloud location, e.g., "us-central1"'), // Optional: Location of the reranking model
+});
+
+// Type alias for the options schema
+export type VertexAIRerankerOptions = z.infer<
+  typeof VertexAIRerankerOptionsSchema
+>;
+
+// Define the structure for each individual reranker configuration
+export const VertexRerankerConfigSchema = z.object({
+  model: z.string().optional().describe('Model name for reranking'), // Optional: Model name, defaults to a pre-defined model
+});
+
+export interface VertexRerankerConfig {
+  name?: string;
+  model?: string;
+}
+
+export interface VertexRerankPluginOptions {
+  rerankOptions: VertexRerankerConfig[];
+  projectId: string;
+  location?: string; // Optional: Location of the reranker service
+}
+
+export interface VertexRerankOptions {
+  authClient: GoogleAuth;
+  pluginOptions?: PluginOptions;
+}
+
+/**
+ * Creates Vertex AI rerankers.
+ *
+ * This function returns a list of reranker actions for Vertex AI based on the provided
+ * rerank options and configuration.
+ *
+ * @param {VertexRerankOptions<EmbedderCustomOptions>} params - The parameters for creating the rerankers.
+ * @returns {RerankerAction<z.ZodTypeAny>[]} - An array of reranker actions.
+ */
+export function vertexAiRerankers(
+  params: VertexRerankOptions
+): RerankerAction<z.ZodTypeAny>[] {
+  if (!params.pluginOptions) {
+    throw new Error(
+      'Plugin options are required to create Vertex AI rerankers'
+    );
+  }
+  const pluginOptions = params.pluginOptions;
+  if (!params.pluginOptions.rerankOptions) {
+    return [];
+  }
+
+  const rerankOptions = params.pluginOptions.rerankOptions;
+  const rerankers: RerankerAction<z.ZodTypeAny>[] = [];
+
+  if (!rerankOptions || rerankOptions.length === 0) {
+    return rerankers;
+  }
+
+  for (const rerankOption of rerankOptions) {
+    const reranker = defineReranker(
+      {
+        name: `vertexai/${rerankOption.name || rerankOption.model}`,
+        configSchema: VertexAIRerankerOptionsSchema.optional(),
+      },
+      async (query, documents, _options) => {
+        const auth = new GoogleAuth();
```

Comment:
Can these 3 lines be moved outside of defineReranker?

Reply:
yup!